### PR TITLE
HDDS-8503. Exclude nodes are not well considered in SCMContainerPlacementRackAware#chooseNodes

### PR DESCRIPTION
**Pull request by ashishkumar50**
## What changes were proposed in this pull request?

When used and favoured nodes are on same rack then there is possibility that all replicas will place into one rack. 
In this PR we are ensuring that if multiple racks are available then replicas will be placed in at least two racks, even though favoured node is on same rack as used node.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8503

## How was this patch tested?

New Unit test.

### CODE REVIEW

**Comment by umamaheswararao** [comment]
@swamirishi could you please help to review it?

**Comment by ashishkumar50** [comment]
@ChenSammi @siddhantsangwan, Can you please help to review.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java [hunk 1]
```diff
@@ -561,7 +561,7 @@ private List<DatanodeDetails> chooseNodes(List<DatanodeDetails> excludedNodes,
       DatanodeDetails favoredNode = favoredNodeNum > favorIndex ?
           favoredNodes.get(favorIndex) : null;
       DatanodeDetails chosenNode;
-      if (favoredNode != null && networkTopology.isSameParent(
+      if (favoredNode != null && !networkTopology.isSameParent(
```
Comment (swamirishi) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java:564]: I guess usedNodes also needs to be considered here & not just excludedNodes & chosenNodes

**Review by swamirishi — CHANGES_REQUESTED** [review]
@ashishkumar50 thanks for the patch. Need some more info about the patch.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java (hunk 1 above)
Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java:564]: @swamirishi Thanks for the review.
UsedNodes are already added into [excludedNodes](https://github.com/apache/ozone/blob/b3c84845926fb79e4b3b4ae20ded2be1c97ada06/hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackAware.java#L227) in the caller. And nodes here we are choosing in different rack than excluded nodes. So it's considering usedNodes as well here.

File: hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/TestSCMContainerPlacementRackAware.java [hunk 2]
```diff
@@ -797,4 +797,44 @@ public void testNoFallbackWithUsedNodes(int datanodeCount) {
...
+    // 1 favoured node
+    favouredNodes.add(datanodes.get(2));
+
+    List<DatanodeDetails> datanodeDetails = policy.chooseDatanodes(usedNodes,
+        null, favouredNodes, nodeNum, 0, 5);
+
+    Assertions.assertEquals(nodeNum, datanodeDetails.size());
+    // Favoured node should not be returned,
+    // as favoured node is in the same rack as both used nodes.
+    Assertions.assertFalse(favouredNodes.get(0).getUuid() ==
```
Comment (ChenSammi) [inline hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/TestSCMContainerPlacementRackAware.java:822]: Can it change to check that the returned node is not under the same rack as favouredNode?

Comment (ashishkumar50) [reply hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/TestSCMContainerPlacementRackAware.java]: @ChenSammi, Thanks for the review. I have updated test to check different rack.

**Review by ChenSammi — APPROVED** [review]

**Pull request merged**